stm32mp1: PSCI and few platform local changes

etienne-lms opened the pull request:
**edited:** shared resources part removed: this P-R is a collection of incremental internal changes in the platform to prepares drivers and shared resource support.

~~This P-R changes paltform stm32mp1 to introduce a shared resource driver Shared resources relate to resources that can be restricted to secure accesses or open to non secure accesses or even be shared by both secure and non secure world through software services implemented in secure world.~~

This change set introduces this support and few platform changes needed to prepare introduction of stm32mp1 drivers and features that will rely of shared resource. As part of such changes, a bit a power support is added, moving previous platform PSCI support into a dedicate source file.

jforissier commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_rcc.c:
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Not a strong opinion, but this driver would be fine in the usual drivers directory I think.

etienne-lms replied:
Few drivers will some be introduced for this platform. Most of them will land in **core/drivers/**, all those that could be platform agnostic. For the few drivers that are really specific to the platform, I located their source in **plat-stm32mp1/drivers/**.

Being in the platform subdir allows to use some short cuts. I.e using a short `RCC_` prefix instead of a larger `STM32MP1_RCC_` in many macros because it should be nicely exported. Also some specific calls will link such platform specific drivers together (shared resources, secure cfg hardening) and I though they were better grouped together this way.

jforissier replied:
ok

jforissier commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_rcc.c:
```diff
@@ -0,0 +1,24 @@
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+
+uintptr_t stm32_rcc_base(void)
```
This pattern is really appearing too many times I think (in this patchset but also more generally in `plat-*/*.c`).
Could `io_pa_or_va()` be used instead? It would have to be moved from `include/drivers/serial.h` to a more generic file, `include/drivers/io.h` maybe?

etienne-lms replied:
An issue is that some device may be secure or non secure and accessing the device should be done according to its secure state. So we must carefully select the appropriate mapping: non secure when the IP is non secure, and secure when it is secure. The scheme used is the get the device state from the DT (status= / secure-status=) and to setup the device driver accordingly.

etienne-lms replied:
Ok, I moved to `phys_to_virt_io()` and will see how `io_pa_or_va()` can be moved to a more generic component.

jforissier commented on core/arch/arm/plat-stm32mp1/pm/psci.c:
```diff
@@ -167,6 +176,31 @@ int psci_cpu_on(uint32_t core_id, uint32_t entry, uint32_t context_id)
+	uint32_t excep;
+
+	if (pos == 0) {
+		EMSG("PSCI_CPU_OFF not supported for core0, use system_off");
```
`s/core0/core #0/`
`s/system_off/PSCI_CPU_OFF/`

Just curious, is this true even if other CPUs are still running?

etienne-lms replied:
Actually the chosen scheme for power sequences expect boot core (core #0) is the one that boots up first and is powered off last (system_off, system_suspend). When secondary boot core shall be deactivated, we expect a call to `psci_cpu_off()` and the core will be reset until being woken back, which cannot be done for core #0.

etienne-lms commented on core/arch/arm/plat-stm32mp1/main.c:
```diff
@@ -73,11 +74,19 @@ void main_init_gic(void)
 	gic_init(&gic_data, (vaddr_t)gicc_base, (vaddr_t)gicd_base);
 	itr_init(&gic_data.chip);
+
+#ifdef CFG_PSCI_ARM32
```
ok, understood.

etienne-lms commented on core/arch/arm/plat-stm32mp1/stm32_util.h:
```diff
@@ -12,4 +12,14 @@
+ * lock is NULL.
+ */
+uint32_t may_spin_lock(unsigned int *lock);
+void may_spin_unlock(unsigned int *lock, uint32_t exceptions);
```
For the NULL lock case, I agree with you. I'll remove the comment and fix the routines.

About MMU: the cores of this platform do not handle exclusive access on the memory interfaces: no cache, no exclusive store instruction. I think it is a somewhat common setup.

jforissier replied:
So, when MMU is disabled, what ensures that there will be no race condition?

etienne-lms replied:
If MMU is disabled, caller is responsible for its sequence. The very few cases where this is used with MMU off can be a specific power sequence where only one core is active, the other being off.

Does that hurts you to have a driver using such `may_spinlock()`? An alternative can be to discard to `may_spinlock()` and let caller check itself the MMU/cache state to take the lock or not.

jforissier replied:
OK that's fine. I just wanted to make sure there could be no race condition. In fact I did not realize there are only 2 cores, it certainly makes things simpler :-)

Testing the MMU state in `may_spinlock()` is fine, there is no reason to put the burden on the caller.

etienne-lms commented:
@jforissier, i propose to keep only the 1st commit in this P-R. I can open one for the stm32mp1 part of these changes. agree? (in case you've not started to review the later commits).
Thanks Travis to recall me I should run checkpatch prior submitting :(

jenswi-linaro commented:
On Wed, Nov 14, 2018 at 8:37 AM, Etienne Carriere
<notifications@github.com> wrote:
>
> @jforissier, i propose to keep only the 1st commit in this P-R. I can open one for the stm32mp1 part of these changes. agree? (in case you've not started to review the later commits).
> Thanks Travis to recall me I should run checkpatch prior submitting :(

Please do, it's so much easier to review like that.

etienne-lms commented:
All comments addressed.

etienne-lms commented:
Hi, I have removed the shared resources part from the P-R (i will update the P-R string) that was not yet reviewed. I will submit a specific P-R.
I have rebased the other changes and squashed the review fixes.

etienne-lms commented:
Merge conflict resolved.
@jforissier, I have removed the shared resource part so that it is easier to review. I will push it in a later P-R.

etienne-lms commented:
tags applied

jforissier commented:
The IBART failure is unrelated to the PR so it should not prevent this from being merged.

jbech-linaro commented:
> The IBART failure is unrelated to the PR so it should not prevent this from being merged.

Indeed, since last run I've enabled yet another temporary workaround for IBART, where I checkout latest on TF-A and apply the patch in https://github.com/ARM-software/arm-trusted-firmware/pull/1724. With that it seems like things are back on track for now (at least #2709 and #2710 and my test https://github.com/jbech-linaro/optee_os/pull/3 seems to work with this temporary hack). I cross my fingers that there are no other issues and that this will be good enough until the TF-A patch has been merged.
